### PR TITLE
Updates to use wrap package

### PR DESCRIPTION
**Pull request by varunagrawal**
Once borglab/wrap#11 lands, these changes will enable the use of the updated wrap module.

We just need to update the wrap subtree in GTSAM before merging this.

### CODE REVIEW

**Comment by varunagrawal** [comment]
@dellaert @ProfFan this is ready for review! :tada:

File: cmake/HandlePython.cmake [hunk 1]
```diff
@@ -1,22 +1,48 @@
...
-    if(${GTSAM_PYTHON_VERSION} STREQUAL "Default")
-        # Get info about the Python3 interpreter
-        # https://cmake.org/cmake/help/latest/module/FindPython3.html#module:FindPython3
-        find_package(Python3 COMPONENTS Interpreter Development)
+  if(${GTSAM_PYTHON_VERSION} STREQUAL "Default")
 
-        if(NOT ${Python3_FOUND})
-            message(FATAL_ERROR "Cannot find Python3 interpreter. Please install Python >= 3.6.")
-        endif()
+    if(${CMAKE_VERSION} VERSION_LESS "3.12.0")
```
Comment (ProfFan) [inline cmake/HandlePython.cmake:5]: Did you test this with 3.9 and 3.12? If so ignore :)

File: wrap/.gitignore [hunk 2]
```diff
@@ -1,2 +1,5 @@
 __pycache__/
 .vscode/
+*build*
+*.egg-info
+cmake/PybindWrap.cmake
```
Comment (ProfFan) [inline wrap/.gitignore:5]: Hmm, why is this file here?

File: wrap/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,52 @@
+cmake_minimum_required(VERSION 3.9)
```
Comment (ProfFan) [inline wrap/CMakeLists.txt:1]: So now it works even without installing, correct?

I mean, this is not installed along with GTSAM, or it is? Maybe we need more testing if it will go with every GTSAM installation...

File: wrap/README.md [hunk 4]
```diff
@@ -8,9 +8,40 @@ It was designed to be more general than just wrapping GTSAM. For notes on creati
...
+```sh
+mkdir build && cd build
+cmake ..
+make install # use sudo if needed
+```
+
+Using `wrap` in your project is straightforward from here. In you `CMakeLists.txt` file, you just need to add the following:
+
+```cmake
+include(PybindWrap)
```
Comment (ProfFan) [inline wrap/README.md:29]: Again, if this goes with every GTSAM then probably we will need to change the name to be `GtwrapPybindWrap` or similar.

File: wrap/cmake/PybindWrap.cmake.in [hunk 5]
```diff
@@ -89,9 +64,9 @@ function(pybind_wrap
   # ~~~
   add_custom_command(OUTPUT ${generated_cpp}
                      DEPENDS ${interface_header}
-                             ${CMAKE_SOURCE_DIR}/wrap/interface_parser.py
-                             ${CMAKE_SOURCE_DIR}/wrap/pybind_wrapper.py
-                             ${CMAKE_SOURCE_DIR}/wrap/template_instantiator.py
+                             @GTWRAP_SOURCE_DIR@/gtwrap/interface_parser.py
+                             @GTWRAP_SOURCE_DIR@/gtwrap/pybind_wrapper.py
```
Comment (ProfFan) [inline wrap/cmake/PybindWrap.cmake.in:68]: This dynamically generates `PybindWrap.cmake`? Wonder why this is necessary. Can we create this without dynamic codegen?

**Review by ProfFan — COMMENTED** [review]
Looks great! Though I am not sure if we should land this fast. More testing is needed to avoid influx of open issues...

File: cmake/HandlePython.cmake (hunk 1 above)
Comment (varunagrawal) [reply cmake/HandlePython.cmake:5]: I tested it with 3.10 which is what comes standard in Ubuntu 20.04.

File: wrap/.gitignore (hunk 2 above)
Comment (varunagrawal) [reply wrap/.gitignore:5]: This PR introduces a `cmake/PybindWrap.cmake.in` file so ignoring the generated file. :slightly_smiling_face:

File: wrap/CMakeLists.txt (hunk 3 above)
Comment (varunagrawal) [reply wrap/CMakeLists.txt:1]: For GTSAM yes, the behavior is the same as the old one. If we follow the standard cmake process for GTSAM, everything should be the same.

The cool thing is that one can also install `wrap` separately now in case they have a project that doesn't need GTSAM.

File: cmake/HandlePython.cmake (hunk 1 above)
Comment (ProfFan) [reply cmake/HandlePython.cmake:5]: Great

File: wrap/CMakeLists.txt (hunk 3 above)
Comment (ProfFan) [reply wrap/CMakeLists.txt:1]: Cool!

File: wrap/cmake/PybindWrap.cmake.in (hunk 5 above)
Comment (varunagrawal) [reply wrap/cmake/PybindWrap.cmake.in:68]: Needed it so that it picks up the correct directory where the `gtwrap` repo is installed in case a project is being wrapped independently.

File: wrap/README.md (hunk 4 above)
Comment (varunagrawal) [reply wrap/README.md:29]: I agree

File: wrap/cmake/PybindWrap.cmake.in (hunk 5 above)
Comment (ProfFan) [reply wrap/cmake/PybindWrap.cmake.in:68]: I'll approve but this is kinda scary for newcomers, maybe a bit more documentation on how this process is carried out?

**Review by ProfFan — APPROVED** [review]

File: wrap/README.md (hunk 4 above)
Comment (varunagrawal) [reply wrap/README.md:29]: This should actually be fine since it installs under the `gtwrap` directory for cmake.

File: wrap/cmake/PybindWrap.cmake.in (hunk 5 above)
Comment (varunagrawal) [reply wrap/cmake/PybindWrap.cmake.in:68]: No more dynamics codegen! :tada:

File: .github/workflows/build-python.yml [hunk 6]
```diff
@@ -12,6 +12,8 @@ jobs:
       CTEST_PARALLEL_LEVEL: 2
       CMAKE_BUILD_TYPE: ${{ matrix.build_type }}
       PYTHON_VERSION: ${{ matrix.python_version }}
+      PYTHON_MAJOR_VERSION: 3
```
Comment (ProfFan) [inline .github/workflows/build-python.yml:15]: Why two variables? Also here note that PYTHON_VERSION is a variable in the build matrix.

Comment (varunagrawal) [reply .github/workflows/build-python.yml:15]: Setting a more specific python version for CMake seemed like a better option, and helps us debug the lowest supported version in the future (helped me figure out the MANIFEST.in issue).

Problem is, that env variable is also used to set the Python interpreter, so it leads to `python3.6.9` which is invalid, hence I added the `PYTHON_MAJOR_VERSION` flag.

File: wrap/CMakeLists.txt [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+# ##############################################################################
+# Install the Python package
+find_package(
+  Python ${WRAP_PYTHON_VERSION}
+  COMPONENTS Interpreter
+  REQUIRED)
+
+# Detect virtualenv and set Pip args accordingly
+# https://www.scivision.dev/cmake-install-python-package/
+if(DEFINED ENV{VIRTUAL_ENV} OR DEFINED ENV{CONDA_PREFIX})
```
Comment (ProfFan) [inline wrap/CMakeLists.txt:46]: Why is this in CMakeLists? TBH nothing in CMakeLists should ever make changes to the system. Anything that modifies the environment should be at install time (`make install`).

Comment (varunagrawal) [reply wrap/CMakeLists.txt:46]: Unfortunately there is no way to around this because the wrapping is done during `make`, and the gtwrap package is required to be installed at that point.

This is one of those instances where we can't quite follow the standard conventions to achieve the end result.

Comment (ProfFan) [reply wrap/CMakeLists.txt:46]: I still think this is probably not what @dellaert wants, but nvm if it is :)

Also, I don't really see the reason `gtwrap` needs to be installed to be used, we can just keep it in the tree.

Comment (varunagrawal) [reply wrap/CMakeLists.txt:46]: If we don't install it, then we get all sorts of import errors in Python. This is a major failing of Python that imports and relative imports don't work as specified.

Comment (varunagrawal) [reply wrap/CMakeLists.txt:46]: > I still think this is probably not what @dellaert wants, but nvm if it is :)

I don't understand this. The idea is that we set up the install of gtwrap as a prerequisite so that at make time, the system has everything it needs to finish successfully. 🙂

Comment (ProfFan) [reply wrap/CMakeLists.txt:46]: Oh, I got what you mean, but I think if we need to install `gtwrap`, then it should be self-contained (a.k.a. installed to a fresh venv in the build folder) and not touch the current user environment.

Comment (varunagrawal) [reply wrap/CMakeLists.txt:46]: Yeah that seemed like more work than necessary TBH. It's just a small package that doesn't have any collisions or messes with the standard library so it seemed fine to me, especially as this vastly simplifies the process on the user's end.

**Comment by varunagrawal** [comment]
@dellaert you will need to do the following since one of the checks has errored out: "You can remove the outdated check in the protected branch settings of the repository settings."

@ProfFan please can you review again?

**Comment by ProfFan** [comment]
Why is the clang test removed?

**Comment by varunagrawal** [comment]
It wasn't. I googled the issue and it seems to be a bug in Actions.

**Comment by varunagrawal** [comment]
@ProfFan since Frank is busy, I am closing this and creating a new PR.
